### PR TITLE
feat(db): add clickhouse

### LINKED ISSUE #27: Tracker logic
## Context

The consumer now processes all events and turns them into `validator_metadata_events` when it thinks there are validators attached to a beacon node. Sentries will try to redo succesful handshakes & metadata exchanges every epoch, which means that in the ideal case we have data points on each beacon node every epoch.

The next step is to build a view on live validators based on these events. The table will look like this (stored in sqlite DB inside consumer)
- PeerID
- ENR
- Multiaddr
- IP
- Port
- LastSeen
- LastEpoch
- **PossibleValidator: bool** 
- **AverageValidatorCount: int**
- **NumObservations: counter**

**Bold properties** are ones that we have to derive. The rest can just be extracted from the metadata events. Note that we maintain a single row per PeerID and PeerID can thus be the primary key.

## Methodology
For each `metadata_event` that results in a `validator_metadata_event`, we do the following:

1. If `shortLivedSubnets > 0` we set `PossibleValidator = true`
2. Increment `NumObservations`
3. Derive estimated number of validators: `AverageValidatorCount` will be a cumulative average of all the validator count estimations so far. The `ValidatorCount` entry for the current event will be `shortLivedSubnets` / 2, because validators at any point will be subscribed to 2 attnets: one for the current epoch and once for the next. NOTE: we should round up not down when doing integer division! I.e. `q := 1 + (x - 1) / y`
4. Set the new `AverageValidatorCount` to be the cumulative average of the previous ones: 
$$new average=current average+\frac{new count}{(new value−current average)}$$
5. Insert or modify row

### LINKED ISSUE #28: Consumer validator REST API for dashboards
We want to have a REST API to get the real-time validator data for the dashboard analytics

### PR DESCRIPTION
**Pull request by namn-grg**
fixes #27 

- Add _optional_ Clickhouse support for sink for storing `validator_metadata`
- Configurable consumer and DB options via cmd

fixes #28 
- Add GET API `/validators` to get all the data from the sqlite database

### CODE REVIEW

File: clickhouse/client.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+		last_seen UInt64,
+		last_epoch UInt64,
+		possible_validator Bool,
+		average_validator_count Int32,
+		num_observations UInt64
+	) ENGINE = MergeTree()
+PRIMARY KEY (peer_id, last_seen)`, db)
+}
+
+type ValidatorMetadataEvent struct {
```
Comment (namn-grg) [inline clickhouse/client.go]: made a seperate type, for now. We can also merge it with `ParquetValidatorEvent` if we want them to have same fields.

File: clickhouse/client.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		c.log.Error().Err(err).Msg("creating validator_metadata table")
+		return err
+	}
+
+	go c.validatorEventBatcher()
+
+	return nil
+}
+
+func (c *ClickhouseClient) validatorEventBatcher() {
```
Comment (namn-grg) [inline clickhouse/client.go:103]: Implementation inspiration was taken from fiber indexer

File: cmd/cmd.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+import (
+	"github.com/chainbound/valtrack/clickhouse"
+	"github.com/chainbound/valtrack/consumer"
+
+	"github.com/rs/zerolog"
+	"github.com/urfave/cli/v2"
+)
+
+var ConsumerCommand = &cli.Command{
```
Comment (namn-grg) [inline cmd/cmd.go:18]: Made a separate file for cleaner and better code

File: consumer/consumer.go [hunk 4]
```diff
@@ -137,113 +153,135 @@ func RunConsumer(natsURL string) {
...
+		Endpoint: cfg.ChCfg.Endpoint,
+		DB:       cfg.ChCfg.DB,
+		Username: cfg.ChCfg.Username,
+		Password: cfg.ChCfg.Password,
+
+		MaxValidatorBatchSize: cfg.ChCfg.MaxValidatorBatchSize,
+	}
+
+	var chClient *ch.ClickhouseClient
+	if chCfg.Endpoint != "" {
```
Comment (namn-grg) [inline consumer/consumer.go:141]: Made clickhouse optional, also operational without it.

File: consumer/consumer.go [hunk 5]
```diff
@@ -137,113 +153,135 @@ func RunConsumer(natsURL string) {
...
+				if err != nil {
+					c.log.Error().Err(err).Msg("Error fetching batch of messages")
+					return
+				}
+				if err = batch.Error(); err != nil {
+					c.log.Error().Err(err).Msg("Error in messages batch")
+					return
+				}
+				for msg := range batch.Messages() {
+					handleMessage(c, msg)
```
Comment (namn-grg) [inline consumer/consumer.go]: Refactored to optimize

File: consumer/consumer.go [hunk 6]
```diff
@@ -327,3 +365,87 @@ func storeMetadataReceivedEvent(pw *writer.ParquetWriter, event ethereum.Metadat
...
+			if err != nil {
+				c.log.Error().Err(err).Msg("Invalid port in multiaddr")
+				continue
+			}
+
+			port, err := strconv.Atoi(portStr)
+			if err != nil {
+				c.log.Error().Err(err).Msg("Invalid port number")
+				continue
+			}
```
Comment (namn-grg) [inline consumer/consumer.go]: It seems not the best way to do it, but it works for now. IP and Port can be added as fields to the NATS event.

File: consumer/consumer.go [hunk 7]
```diff
@@ -327,3 +365,87 @@ func storeMetadataReceivedEvent(pw *writer.ParquetWriter, event ethereum.Metadat
...
+			c.validatorCache[event.ID] = &validatorMetadata
+
+			// Write to Clickhouse
+			if c.chClient != nil {
+				c.chClient.ValidatorEventChan <- &validatorMetadata
+				c.log.Trace().Any("validator_metadata", validatorMetadata).Msg("Inserted validator metadata")
+			}
+		default:
+			c.log.Debug().Msg("No validator metadata event")
+			time.Sleep(1 * time.Second) // Prevents busy waiting
```
Comment (namn-grg) [inline consumer/consumer.go]: Read about this - busy waiting leads to more CPU utilization and energy consumption. (Not very useful here, can be in big scalable software)

File: consumer/utils.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	for _, i := range slice {
+		if i == item {
+			return true
+		}
+	}
+	return false
+}
+
+func ComputeNewAvg(prevAvg int32, prevCount uint64, currValidatorCount int) int32 {
+	return int32((int64(prevCount)*int64(prevAvg) + int64(currValidatorCount)) / int64(prevCount+1))
```
Comment (namn-grg) [inline consumer/utils.go]: The formula for the new average in the issue seemed incorrect which was. - $$new average=current average+\frac{new count}{(new value−current average)}$$

I think the correct formula (the one implemented here) is -
$$new average=\frac{(current average * current count) + current validator count}{current count +1}$$

Comment (namn-grg) [reply consumer/utils.go]: Note the average is rounded down here

File: clickhouse/client.go (hunk 1 above)
Comment (mempirate) [reply clickhouse/client.go]: Let's have one struct for each event, and then use the tag annotations to make it compatible with multiple sinks (json, parquet, clickhouse).

File: clickhouse/client.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/ClickHouse/clickhouse-go/v2/lib/driver"
+	"github.com/chainbound/valtrack/log"
+	"github.com/rs/zerolog"
+)
+
+func ValidatorMetadataDDL(db string) string {
+	return fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s.validator_metadata (
```
Comment (mempirate) [inline clickhouse/client.go:17]: This should be the table in sqlite, not Clickhouse. Clickhouse only stores the regular events like the ones we've been storing in Parquet

**Review by mempirate — CHANGES_REQUESTED** [review]

File: types/types.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
+package types
```
Comment (namn-grg) [inline types/types.go:1]: Made a different package to store events type.

File: types/types.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	CrawlerLoc string `parquet:"name=crawler_location, type=BYTE_ARRAY, convertedtype=UTF8" json:"crawler_location" ch:"crawler_location"`
+	Timestamp  int64  `parquet:"name=timestamp, type=INT64" json:"timestamp" ch:"timestamp"`
+}
+
+type MetadataReceivedEvent struct {
+	ENR               string          `parquet:"name=enr, type=BYTE_ARRAY, convertedtype=UTF8" json:"enr" ch:"enr"`
+	ID                string          `parquet:"name=id, type=BYTE_ARRAY, convertedtype=UTF8" json:"id" ch:"id"`
+	Multiaddr         string          `parquet:"name=multiaddr, type=BYTE_ARRAY, convertedtype=UTF8" json:"multiaddr" ch:"multiaddr"`
+	Epoch             int             `parquet:"name=epoch, type=INT32" json:"epoch" ch:"epoch"`
+	MetaData          *SimpleMetaData `parquet:"name=metadata, type=BYTE_ARRAY, convertedtype=UTF8" json:"metadata" ch:"metadata"`
```
Comment (namn-grg) [inline types/types.go:40]: Changed it from `eth.MetaDataV1` to  custom `SimpleMetaData` to support parquet in single type

File: types/types.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+type ValidatorEvent struct {
+	ENR       string `parquet:"name=enr, type=BYTE_ARRAY, convertedtype=UTF8" json:"enr" ch:"enr"`
+	ID        string `parquet:"name=id, type=BYTE_ARRAY, convertedtype=UTF8" json:"id" ch:"id"`
+	Multiaddr string `parquet:"name=multiaddr, type=BYTE_ARRAY, convertedtype=UTF8" json:"multiaddr" ch:"multiaddr"`
+	Epoch     int    `parquet:"name=epoch, type=INT32" json:"epoch" ch:"epoch"`
+	// MetaData          *SimpleMetaData `parquet:"name=metadata, type=BYTE_ARRAY, convertedtype=UTF8" json:"metadata" ch:"metadata"`
+	SeqNumber         int64   `parquet:"name=seq_number, type=INT64" json:"seq_number" ch:"seq_number"`
+	Attnets           string  `parquet:"name=attnets, type=BYTE_ARRAY, convertedtype=UTF8" json:"attnets" ch:"attnets"`
+	Syncnets          string  `parquet:"name=syncnets, type=BYTE_ARRAY, convertedtype=UTF8" json:"syncnets" ch:"syncnets"`
```
Comment (namn-grg) [inline types/types.go]: Clickhouse doesnt support struct afaik so made it into separate fields

File: types/types.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
+package types
+
+type ValidatorEvent struct {
```
Comment (namn-grg) [inline types/types.go:5]: This event gets published to clickhouse

File: clickhouse/client.go (hunk 9 above)
Comment (namn-grg) [reply clickhouse/client.go:17]: Made the necessary changes! 
- The `ValidatorEvent` is published to clickhouse
- The validator tracker table is stored on the local sqlite database named `validator_tracker.sqlite`

File: consumer/db.go [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+	"strconv"
+	"time"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+var (
+	createTableQuery = `
+    CREATE TABLE IF NOT EXISTS validator_tracker (
+        PeerID TEXT PRIMARY KEY,
```
Comment (mempirate) [inline consumer/db.go]: SQL columns use `snake_case` naming conventions, let's change it!

**Review by mempirate — CHANGES_REQUESTED** [review]
LGTM, looking good and I'm gonna try to run it now, requested slight nit on SQL column naming

File: consumer/db.go [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        ENR TEXT,
+        Multiaddr TEXT,
+        IP TEXT,
+        Port INTEGER,
+        LastSeen TEXT,
```
Comment (mempirate) [inline consumer/db.go]: This can be INTEGER

**Pull request merged**